### PR TITLE
WIP: Status backend intergation

### PR DESCRIPTION
**Pull request by ulisesmac**
This PR is to show the current progress on the status-backend integration.

## Setup

If you are willing to run this PR, these are the steps needed:

1. Run status-backend at <PORT>
2. Run an Android emulator or physical device and reverse the <PORT>: 
     `adb reverse tcp:<PORT> tcp:<PORT>`
     This is needed to communicate the app with the back-end running on your computer
3. In Status mobile, go to the `ns` `status-backend.config` and define `data-dir-path` with your own.
4. Run the app as always.

If the app is stuck in a black screen, it means we have no connection with status-backend. Sometimes it crashes while creating a profile, sometimes after those steps. **This still is a WIP**:

https://github.com/user-attachments/assets/d6645e65-dff2-4763-ba98-7c8ae8cd5e44

## Implementation details

Please refer to the ns `status-backend.core` for the main implementation and some REPL commands to play with.

## Strategy to follow

In order to completely integrate status-backend, we need to port some native modules, [see the discussion here](https://discord.com/channels/1210237582470807632/1273675588846096446/1299122526160551967). 

So this PR is presenting an initial progress that consist of:

- :white_check_mark:  Web socket based signals are 100% working
-  :white_check_mark:  RPC calls are 100% working
- :exclamation:  A general mechanism to invoke status back-end methods async. This step needs to manually port some native modules since we expect a sync behavior, they will be addressed in following PRs

### CODE REVIEW

File: src/utils/network/core.cljs [hunk 1]
```diff
@@ -41,7 +41,12 @@
        (map :error)
        (not-any? identity)))
 
-(defn chain-id-available?
-  [current-networks network]
-  (let [chain-id (get-in network [:config :NetworkId])]
-    (every? #(not= chain-id (get-in % [1 :config :NetworkId])) current-networks)))
```
Comment (ulisesmac) [inline src/utils/network/core.cljs:47]: NOT RELATED: Removed since this namespace is not related to wallet stuff, and also it has 0 usages

File: src/utils/network/core.cljs [hunk 2]
```diff
@@ -41,7 +41,12 @@
...
+(defn fetch [url {:keys [body] :as params} callback]
+  (let [js-params (cond-> params
+                    (map? body) (update :body (comp js/JSON.stringify clj->js))
+                    :always     clj->js)]
+    (-> (js/fetch url js-params)
+        (.then (fn [response]
+                 (.text response)))
+        (.then callback)
+        (.catch #(log/error (str "Error while fetching: " url) %)))))
```
Comment (ulisesmac) [inline src/utils/network/core.cljs]: Added a general `fetch` function that is wrapping `js/fetch` to use it in a Clojure friendly way

File: src/utils/re_frame.cljs [hunk 3]
```diff
@@ -21,8 +21,10 @@
              (when js/goog.DEBUG
                (reset! handler-nesting-level 0))
              (when rf.interop/debug-enabled?
-               (log/debug "Handling re-frame event: "
-                          (first (interceptor/get-coeffect context :event))))
+               (do
+                 #_(log/debug "Handling re-frame event: "
+                            (first (interceptor/get-coeffect context :event)))
```
Comment (ulisesmac) [inline src/utils/re_frame.cljs]: Commented some log calls, this'll be reverted. They are being too noisy to test the back-end

File: src/status_im/core.cljs [hunk 4]
```diff
@@ -40,7 +40,7 @@
 
 (def adjust-resize 16)
 
-(defn is-hermes
+(defn hermes?
```
Comment (ulisesmac) [inline src/status_im/core.cljs]: Again, not related, couldn't resist, renamed this function because this is not JS

File: src/status_backend/core.cljs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  (oops/oset! web-socket "onopen" #(log/debug "[backend] Web Socket connected"))
+  (oops/oset! web-socket "onmessage" #(-> % (oops/oget "data") (on-message)))
+  (oops/oset! web-socket "onerror" #(log/error "[backend] Web Socket error" %))
+  (oops/oset! web-socket "onclose" #(log/error "[backend] Web Socket closed" %)))
+
+
+(defn call-private-rpc [payload callback]
+  ;; NOTE: according to the status-backend readme, we call RPCs using the `CallRPC`
+  ;; endpoint, instead of using `CallPrivateRPC` as in mobile.
+  (fetch "CallRPC" payload callback))
```
Comment (ulisesmac) [inline src/status_backend/core.cljs]: According to status backend readme, we should invoke `CallRPC` for the services, but in mobile we use the effect `:json-rpc/call` which wraps `CallPrivateRPC`.

Why do we changed this? what is the difference between `CallRPC` and `CallPrivateRPC`? :thinking: 

cc: @ilmotta @qfrank

File: src/status_backend/core.cljs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                          (partial fetch endpoint)))}))
+
+(declare web-socket)
+
+(defn init-web-socket [on-message]
+  (defonce web-socket (js/WebSocket. config/signals-url))
+  (oops/oset! web-socket "onopen" #(log/debug "[backend] Web Socket connected"))
+  (oops/oset! web-socket "onmessage" #(-> % (oops/oget "data") (on-message)))
+  (oops/oset! web-socket "onerror" #(log/error "[backend] Web Socket error" %))
+  (oops/oset! web-socket "onclose" #(log/error "[backend] Web Socket closed" %)))
```
Comment (ulisesmac) [inline src/status_backend/core.cljs]: This will handle all the signals

File: src/status_backend/core.cljs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+(def fetch-js-obj
+  "Performs a request to the status-backend (by using `status-backend.core/fetch`) using
+  the method invoked as endpoint name. E.g
+  This call: (.myMethod fetch-js-obj params)
+  Is the same as: (fetch \"MyMethod\" params)
+  "
+  (js/Proxy. #js{}
+             #js{:get (fn [_target native-module-method-name]
+                        (let [endpoint (method-name->endpoint native-module-method-name)]
+                          (partial fetch endpoint)))}))
```
Comment (ulisesmac) [inline src/status_backend/core.cljs]: This will handle ~90% of the native module calls

File: src/status_backend/config.cljs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+(ns status-backend.config)
```
Comment (ulisesmac) [inline src/status_backend/config.cljs]: This file should be simplified with our nix tooling. But currently you need to add your own paths

File: src/native_module/core.cljs [hunk 9]
```diff
@@ -14,51 +16,69 @@
 
 (defn account-manager
   []
-  (when (exists? (.-NativeModules react-native))
-    (.-AccountManager ^js (.-NativeModules react-native))))
+  (if false
+    (when (exists? (.-NativeModules react-native))
+      (.-AccountManager ^js (.-NativeModules react-native)))
+    status-backend/fetch-js-obj))
```
Comment (ulisesmac) [inline src/native_module/core.cljs]: along this file I use `if false` as a  temporary pattern to ignore the original module implementation and use the back-end fetch logic

File: src/status_backend/config.cljs (hunk 8 above)
Comment (ilmotta) [reply src/status_backend/config.cljs]: These settings could be configured using `:clojure-defines` in `shadow-cljs.edn`, with a sensible default being injected by a make target. Similar to how we run CLJS tests with the line `:ns-regexp #shadow/env "SHADOW_NS_REGEXP"`.

The things that seem to require customization are the `address`, `port`, and `data-dir-path`, the rest is all derived. Or is there another setting you think we should parameterize?

File: src/native_module/core.cljs [hunk 10]
```diff
@@ -522,23 +547,37 @@
...
+    status-backend.config/data-dir-path))
 
 (defn fleets
   []
   (.fleets ^js (status)))
 
 (defn keystore-dir
   []
-  (.keystoreDir ^js (utils)))
+  (if false
```
Comment (ilmotta) [inline src/native_module/core.cljs]: These checks could be replaced by something like `when ^boolean js/goog.DEBUG` to allow dead code elimination, but with a different definition from `goog-define` and shadow-cljs, as we do in status-im config.

File: src/native_module/core.cljs [hunk 11]
```diff
@@ -522,23 +547,37 @@
...
+    (.keystoreDir ^js (utils))
+    status-backend.config/keystore-dir-path))
 
 (defn log-file-directory
   []
-  (.logFileDirectory ^js (log-manager)))
+  (if false
```
Comment (ilmotta) [inline src/native_module/core.cljs]: Instead of conditionals being repeated multiple times, we could have two implementations of a protocol, in two different namespaces to avoid any risk of mixing up things. It would be a bit better to have the status-backend as decoupled from production code as possible.

Edit: let's forget about my suggestion 😅 because the problem is that the `native-module.core` namespace is accessed directly by the rest of the code in the repo, which means we would need to change all places to access the correct instance implementing the protocol, and not the function vars directly. But this would require a ton of effort, not worth it, quite risky.

File: src/status_backend/core.cljs [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+(declare web-socket)
+
+(defn init-web-socket [on-message]
+  (defonce web-socket (js/WebSocket. config/signals-url))
+  (oops/oset! web-socket "onopen" #(log/debug "[backend] Web Socket connected"))
+  (oops/oset! web-socket "onmessage" #(-> % (oops/oget "data") (on-message)))
```
Comment (ilmotta) [inline src/status_backend/core.cljs]: Nit: `(on-message)` can be written without the parentheses

File: src/status_backend/core.cljs (hunk 5 above)
Comment (ilmotta) [reply src/status_backend/core.cljs]: The current implementations in status-go are identical:

https://github.com/status-im/status-go/blob/132ea05fc8735b91922a440009efdbc7f1b8b8ce/mobile/status.go#L249-L273

And https://github.com/status-im/status-go/blob/132ea05fc8735b91922a440009efdbc7f1b8b8ce/api/geth_backend.go#L2233-L2249

@igor-sirotin and @qfrank can probably help clarify this point.

File: src/utils/network/core.cljs (hunk 2 above)
Comment (ilmotta) [reply src/utils/network/core.cljs]: I always find confusing to mix promises with callbacks. The `fetch` function is returning a promise, so we don't need a callback. I've only seen this style of code mixing promises with callbacks in status-mobile.

But I imagine you did it like this to avoid further changes in the repo? Which would be caused if you used an implementation like the one below.

```clojure
(defn fetch [url {:keys [body] :as params}]
  (let [js-params (cond-> params
                    (map? body) (update :body (comp js/JSON.stringify clj->js))
                    :always     clj->js)]
    (-> (js/fetch url js-params)
        (.then (fn [response]
                 (.text response)))
        (.catch #(log/error (str "Error while fetching: " url) %)))))
```

I think it's best to keep changes at a minimum, but I wanted to point out @ulisesmac. Sorry if you already know all this.

**Review by ilmotta — COMMENTED** [review]
Good progress @ulisesmac! I did check the branch locally and the app crashes at the same spot you described in the video.

Maybe another CC, such as @qfrank can help you spot the cause of the crash?

File: src/status_backend/core.cljs (hunk 12 above)
Comment (ulisesmac) [reply src/status_backend/core.cljs]: Yeah, I know, it's just a matter of code style preferences I think.

I remove the parentheses when the previous call didn't have it

File: src/status_backend/core.cljs (hunk 5 above)
Comment (qfrank) [reply src/status_backend/core.cljs]: There's no difference after this [PR](https://github.com/status-im/status-go/pull/2275) , we can just use `CallRPC`, if you want to know details about the `private` things, maybe @Samyoul knows something

File: src/status_backend/core.cljs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+(def fetch-js-obj
+  "Performs a request to the status-backend (by using `status-backend.core/fetch`) using
+  the method invoked as endpoint name. E.g
+  This call: (.myMethod fetch-js-obj params)
+  Is the same as: (fetch \"MyMethod\" params)
+  "
+  (js/Proxy. #js{}
+             #js{:get (fn [_target native-module-method-name]
+                        (let [endpoint (method-name->endpoint native-module-method-name)]
```
Comment (qfrank) [inline src/status_backend/core.cljs]: this won't work ... We cannot simply determine the endpoint name based on the `native-module-method-name` because there are many unsupported endpoints in the status-backend server. Therefore, I created this [PR 5943](https://github.com/status-im/status-go/pull/5943) to solve this problem. For some unsupported endpoints, we need to use the corresponding V2 versions. 
I also created another [PR](https://github.com/status-im/status-mobile/pull/21450)(not related to status backend integration) to use the V2 endpoint so that we can remove the old endpoint in the future

**Review by qfrank — COMMENTED** [review]
I saw there're many other methods like `create-account-and-login` / `init-status-go-logging` are not using status-backend server yet, but you will update it later right? Because this is just a solution demo draft?

File: src/status_backend/core.cljs (hunk 13 above)
Comment (qfrank) [reply src/status_backend/core.cljs]: Based on Igor's this [reply](https://discord.com/channels/1210237582470807632/1273675588846096446/1299368407635853385), I still need to supplement some missing V2 endpoints in PR#5943, such as `AppStateChange`. I plan to complete the supplementation today.

Comment (qfrank) [reply src/status_backend/core.cljs]: > Based on Igor's this [reply](https://discord.com/channels/1210237582470807632/1273675588846096446/1299368407635853385), I still need to supplement some missing V2 endpoints in PR#5943, such as `AppStateChange`. I plan to complete the supplementation today.

done

File: src/native_module/core.cljs (hunk 11 above)
Comment (seanstrom) [reply src/native_module/core.cljs]: Maybe we could create to different implementations / protocols that are conditionally required from this namespace? For example,`native-module.core` would conditionally require `native-module.mobile` when in production (or when we want to test the bindings), and another namespace like `native-module.backend` could be required during development. Thoughts?

File: src/native_module/core.cljs [hunk 14]
```diff
@@ -522,23 +547,37 @@
...
-  (.initLogging ^js (log-manager) enable? mobile-system? log-level log-request-go? callback))
+  (if false
+    (.initLogging ^js (log-manager) enable? mobile-system? log-level log-request-go? callback)
+    (.initLogging ^js (log-manager)
+                  {:Enabled        enable?
+                   :MobileSystem   mobile-system?
+                   :Level          log-level
+                   :LogRequestGo   log-request-go?
+                   :LogRequestFile status-backend.config/log-request-file-path}
+                  callback)))
```
Comment (seanstrom) [inline src/native_module/core.cljs]: Why do we call `.initLogging` on `(log-manager)` for both cases here?

Comment (ulisesmac) [reply src/native_module/core.cljs]: the new one calls status-backend in a compatible way

File: src/native_module/core.cljs (hunk 11 above)
Comment (ilmotta) [reply src/native_module/core.cljs]: @seanstrom I remember now you found a way to conditionally require namespaces. That could work well 👍🏼 Just feels hacky, but might be the best.

Usually, in "good" backend Clojure repos, there's usually a controlled way to initialize system dependencies, and while supporting RDD. In frontend repos with CLJS, it's way less common practice. System initialization being different from dependency injection, but some Clojure libs help solve both problems.

File: src/native_module/core.cljs (hunk 14 above)
Comment (seanstrom) [reply src/native_module/core.cljs]: oh does it change the destination of the logs to point to the status backend?

File: src/status_backend/core.cljs (hunk 13 above)
Comment (ulisesmac) [reply src/status_backend/core.cljs]: Hello @qfrank 
Thank you for your review!

> this won't work ... We cannot simply determine the endpoint name based on the native-module-method-name because there are many unsupported endpoints in the status-backend server

I know there are unsupported endpoints, I'm not expecting this draft PR fully works right now, but this could be a pattern to follow.

Regarding to:

> For some unsupported endpoints, we need to use the corresponding V2 versions.
> I also created another https://github.com/status-im/status-mobile/pull/21450(not related to status backend integration) to use the V2 endpoint so that we can remove the old endpoint in the future


I don't fully agree with the approach taken. If we cannot infer the endpoint name based on the native module method name it implies:

1. We need to **manually** maintain a map (native-method -> endpoint) which will be error prone and will require an extra step from everyone while integrating a new native method. Since status-backend won't be used in production, we should reduce the divergence to the minimum.
2. Lack of standarization/consistency: Why do we want to give different names to functions that should only be wrappers?
3. As said in `2`, IMO, we should keep native modules related to status-go as thin as possible, so I don't agree with the approach taken in:
   - https://github.com/status-im/status-mobile/pull/21450
 
   Since it exposes a native method signature that is different from the V2 endpoint called inside. By doing this, we add an extra layer of abstraction/opacity from a call in CLJS to an execution in Status-go. Again, IMO, that PR should update the native methods signature and fix the usages in the CLJS side (i.e. call these native methods with a single param being a JS object).
   

Another option might be to integrate the status-backend calls inside the native module methods, so it'd be implementing the integration in Java/Kotlin, but think that'd be harder to maintain.

CC: @ilmotta

**Comment by ulisesmac** [comment]
@qfrank 

> I saw there're many other methods like create-account-and-login / init-status-go-logging are not using status-backend server yet, but you will update it later right? Because this is just a solution demo draft?

Oh, these methods are already using the status-backend :thinking: the substitution is probably unclear since I wanted to have a low impact in the codebase. We call the status backend when we execute a method from a native module.

Feel free to ping me on Discord if you have more questions.

File: src/utils/network/core.cljs (hunk 2 above)
Comment (ulisesmac) [reply src/utils/network/core.cljs]: Thanks for paying attention to this @ilmotta 

I even considered to return `nil` for this function, so it could be considered just as a function with side effects.

I know it's confusing, but I wanted to pass a callback as a way to easily deal with `fetch`, not having to worry about dealing with a promise, possibbly import  `promesa.core` and propagate the promse (e.g. `status-backend.core/fetch` would also return a promise).

IMO we don't need all the flexibility/power of a promise for `js/fetch`, we just care about having a callback (or two: on-success and on-failure) to receive the result.

I'm open to listening to opinions on why this approach is worse. I wanted to keep things simple and familiar.

File: src/status_backend/config.cljs (hunk 8 above)
Comment (ulisesmac) [reply src/status_backend/config.cljs]: Nope, this is everything. I'll use `:clojure-defines`. Thank you!

File: src/native_module/core.cljs (hunk 10 above)
Comment (ulisesmac) [reply src/native_module/core.cljs]: Yeah, I was thinking that during the implementation, we don't want this code in release builds. I'll implement it, thank you

File: src/native_module/core.cljs (hunk 11 above)
Comment (ulisesmac) [reply src/native_module/core.cljs]: @seanstrom 

What you say sounds like creating a reader conditional to perform the expected require :thinking:

@ilmotta 
I don't like I had to use an `if` here, as I've been saying, I tried to impact as less as possible. I think this should work not so bad.

File: src/status_backend/core.cljs (hunk 13 above)
Comment (ilmotta) [reply src/status_backend/core.cljs]: @ulisesmac, I'll share my perspective, which can be incomplete.

> We need to manually maintain a map (native-method -> endpoint) which will be error prone

I was imagining this manual mapping will only exist temporarily, until status-backend has no unsupported endpoints or very few (which would be easy to handle with a blocklist).

> Lack of standarization/consistency: Why do we want to give different names to functions that should only be wrappers?

Isn't the case that unsupported endpoints always use the equivalent name, but with a `V2` suffix? This is a temporary situation, but there is some consistency.

> As said in 2, IMO, we should keep native modules related to status-go as thin as possible, so I don't agree with the approach taken in:

In the PR https://github.com/status-im/status-mobile/pull/21450, I understood that @qfrank opted to change the native implementation instead of the Clojure layer because  all the changes to build up the JSON request and call the v2 functions would be isolated as internal implementation, and no Clojure code had to be changed in that PR. If, in that PR, we change the native function signatures to take a JSON payload, we will need to change all CLJS calls as well. Maybe it was just a case of different devs with different tech skills and what was most comfortable.

But I have to agree with you @ulisesmac, our agreement is that we only write native code if truly necessary (e.g. performance), otherwise we use Clojure, which is easier to manipulate for most CCs and works well with hot-reload and the REPL.

File: src/native_module/core.cljs (hunk 11 above)
Comment (ilmotta) [reply src/native_module/core.cljs]: It's kind of bad IMO @ulisesmac that we have to sprinkle conditionals in all functions. This mix of dev-only and prod code repeated all the time is a bit error prone. Ideally we would use something like an instance of a record implementing a protocol and completely separate namespaces. Might be doable in the future as an interesting refactoring exercise.

@seanstrom's idea is based on this https://shadow-cljs.github.io/docs/UsersGuide.html#_conditional_reading, which is like a special feature only supported by shadow-cljs.

----

I know I started this discussion, but I think using the `^boolean FLAG` checks in each function is fine for now 👍🏼 The simplest that works

File: src/native_module/core.cljs (hunk 14 above)
Comment (ulisesmac) [reply src/native_module/core.cljs]: Yes.

File: src/status_backend/core.cljs (hunk 13 above)
Comment (qfrank) [reply src/status_backend/core.cljs]: V2 endpoints bring following benefits as it use json data structure:
- it will be easier to return an error 
- standarization/consistency for request/response between frontend/backend
- ignore sensitive fields when log request/response
- don't need to wait adaption from desktop side

I want this feature ASAP as I feel painful with build experience! Perhaps we can simply implement requesting the backend server in a native way that does not require much modification to ClojureScript. Only the configuration (server address/if enabled) needs to be passed to the native code. I can do it based on #21450 , WDYT? @ulisesmac @ilmotta

Comment (ulisesmac) [reply src/status_backend/core.cljs]: @qfrank 

 I agree on having a V2 version with a different API, for sure it is an improvement.
 
What I suggest is:
- Update the native methods' signature to match the V2 endpoints.
- Update the CLJS calls to these native methods (I can help with it, shouldn't be hard).
- Integrate Status backend as suggested in this PR, by just adding a conditional call.

However, if we prefer to have a native implementation then it's fine for me. IMO implementing the request mechanism in Kotlin and another in ObjC is harder to test and maintain. Not sure, but according to the new arch docs, we could (when we adopt it) implement it in C++ and share it across platforms.

**Comment by ulisesmac** [comment]
@ilmotta Pushed new code adding the shadow-cljs config stuff.

If the env var is marked  (or the `goog-define` is set) as `true` the status backend runs. Ofc, it lacks of the proper migration for all the endpoints, we cannot complete the onboarding yet, but I think the code is good enough to be merged.

The only one point that I didn't addresss (since it still has an open discussion) is about the callback added in a promise flow.

File: src/status_backend/core.cljs (hunk 13 above)
Comment (qfrank) [reply src/status_backend/core.cljs]: discard #21450 totally seems not good as we have to keep the old(V1) endpoints in status-go then :thinking:

maybe adjust #21450 and update ReactMethod siganature, e.g.:
```kotlin
@ReactMethod
    fun inputConnectionStringForBootstrapping(connectionString: String, configJSON: String, callback: Callback) {
        val receiverClientConfig = JSONObject(configJSON)
        val params = JSONObject().apply {
            put("connectionString", connectionString)
            put("receiverClientConfig", receiverClientConfig)
        }
        val jsonString = params.toString()
        utils.executeRunnableStatusGoMethod(
            { Statusgo.inputConnectionStringForBootstrappingV2(jsonString) },
            callback
        )
    }
```

to 

```kotlin
@ReactMethod
    fun inputConnectionStringForBootstrappingV2(jsonParam String, callback: Callback) {
        ...
    }
```

 @ulisesmac

Comment (ilmotta) [reply src/status_backend/core.cljs]: > I want this feature ASAP as I feel painful with build experience! Perhaps we can simply implement requesting the backend server in a native way that does not require much modification to ClojureScript. Only the configuration (server address/if enabled) needs to be passed to the native code. I can do it based on https://github.com/status-im/status-mobile/pull/21450 , WDYT? @ulisesmac @ilmotta

@qfrank @ulisesmac This is all highly technical work, with no impact to users, therefore, we can first try to make things work, even if they are kind of dirty, if the code is safe to merge of course. On the other hand, unfortunately every time we change these parts of the code we risk introducing regressions, and the cost of asking a QA to review such PRs is considerable because they have to check a lot of stuff, so let's keep in mind the cost of follow-up PRs touching so many endpoints.

But whatever you guys align I'll probably agree because, in the end, I also just want this integration to work first, so we can all profit from it even during the next 2.32 release.

File: nix/deps/gradle/generate.sh [hunk 15]
```diff
@@ -42,22 +42,10 @@ function gen_deps_list() {
 # https://github.com/status-im/status-mobile/issues/15447
 function add_deps_hack() {
     echo -n \
-'com.android.tools.build:gradle:1.3.1
-com.squareup.okio:okio:1.13.0
-com.squareup.okio:okio:1.15.0
-com.squareup.okhttp3:okhttp:3.12.1
```
Comment (jakubgs) [inline nix/deps/gradle/generate.sh]: Why does this PR include changes from Sid's PR?

Comment (ilmotta) [reply nix/deps/gradle/generate.sh]: There's a rebase issue because the PR now includes a few unrelated commits.

Comment (ulisesmac) [reply nix/deps/gradle/generate.sh]: not sure why, but I've fixed it

**Review by qfrank — CHANGES_REQUESTED** [review]
Pls use this [PR](https://github.com/status-im/status-mobile/pull/21550) instead :) @ulisesmac @ilmotta

**Comment by ulisesmac** [comment]
We changed the approach to a native implementation
